pymc/distributions/transforms.py

```diff
@@ -138,6 +140,115 @@ def log_jac_det(self, value, *inputs):
         return pt.sum(y, axis=-1)
 
 
+class CholeskyCorr(Transform):
+    """
+    Transforms the off-diagonal elements of a correlation matrix to
+    unconstrained real numbers.
+
+    Note: This is not particular to the LKJ distribution - it is only a
+    transform to help generate cholesky decompositions for random valid
+    correlation matrices.
+
+    Ported from here: https://github.com/tensorflow/probability/blob/94f592af363e13391858b48f785eb4c250912904/tensorflow_probability/python/bijectors/correlation_cholesky.py#L31
+
+    The backward side of this transformation is the off-diagonal upper
+    triangular elements of a correlation matrix, specified in row major order.
+    """
+
+    name = "cholesky-corr"
+
+    def __init__(self, n):
+        """
+
+        Parameters
+        ----------
+        n: int
+            Size of correlation matrix
+        """
+        self.n = n
+        self.m = int(n*(n-1)/2) # number of off-diagonal elements
+        self.tril_r_idxs, self.tril_c_idxs = self._generate_tril_indices()
+        self.triu_r_idxs, self.triu_c_idxs = self._generate_triu_indices()
+
+    def _generate_tril_indices(self):
+        row_indices, col_indices = np.tril_indices(self.n, -1)
+        return (
+            pytensor.shared(row_indices),
+            pytensor.shared(col_indices)
+        )
+
+    def _generate_triu_indices(self):
+        row_indices, col_indices = np.triu_indices(self.n, 1)
+        return (
+            pytensor.shared(row_indices),
+            pytensor.shared(col_indices)
+        )
+
+    def _jacobian(self, value, *inputs):
+        return pt.jacobian(
+            self.backward(value),
+            wrt=value
+        )
+
+    def log_jac_det(self, value, *inputs):
+        """
+        Compute log of the determinant of the jacobian.
+
+        There are no clever tricks here - we literally compute the jacobian
+        then compute its determinant then take log.
+        """
+        jac = self._jacobian(value)
+        return pt.log(pt.linalg.det(jac))
+
+    def forward(self, value, *inputs):
+        """
+        Convert the off-diagonal elements of a cholesky decomposition of a
+        correlation matrix to unconstrained real numbers.
+        """
+        # The correlation matrix is specified via its upper triangular elements
+        corr = pt.set_subtensor(
+            pt.zeros((self.n, self.n))[self.triu_r_idxs, self.triu_c_idxs],
+            value
+        )
+        corr = corr + corr.T + pt.eye(self.n)
+
+        chol = pt.linalg.cholesky(corr)
+
+        # Are the diagonals always guaranteed to be positive?
+        # I don't know, so we'll use abs
+        row_norms = 1/pt.abs(pt.diag(chol))
+
+        # Multiply by the row norms to undo the normalization
+        unconstrained = chol*row_norms[:, pt.newaxis]
+
+        return unconstrained[self.tril_r_idxs, self.tril_c_idxs]
+
+    def backward(self, value, *inputs, foo=False):
```

Comment:
You need to check that these functions match the expected outputs from TFP. I used the test case from the tfp docs and got the wrong values -- `array([0.89442719, 0.81649658, 0.91287093])` vs the reference solution

```
array([[1.        , 0.        , 0.        ],
       [0.70710678, 0.70710678, 0.        ],
       [0.66666667, 0.66666667, 0.33333333]])
```

You did some extra research so I might be missing something?

Reply:
Something like this matches tfp:

```
    def backward(self, value, *inputs):
        """
        Convert unconstrained real numbers to the off-diagonal elements of the
        cholesky decomposition of a correlation matrix.
        """
        def unpack_upper_tril_with_eye_diag(x, core_shape):
            """1D allocation case"""
            return pt.set_subtensor(pt.eye(core_shape)[*np.tril_indices(core_shape, k=-1)], x[::-1])
        
        value = pt.as_tensor_variable(value)
        core_shape = value.type.shape[-1]
        
        # Vectorize the 1D case to handle potential batch dimensions
        out = pt.vectorize(partial(unpack_upper_tril_with_eye_diag, core_shape=core_shape), '(n)->(n,n)')(value)
        
        # Vector L2 norm without .real call to speed things up a bit
        norm = pt.sqrt(pt.sum((out ** 2), axis=-1, keepdims=True))
        return out / norm
```

Reply:
Thanks for that code above, there's some great stuff I can reuse.

Need to address this comment first, since actually working in the first place is really fundamental. Here's a notebook that demonstrates that this implementation does replicate the original reference implementation from TFP: https://colab.research.google.com/drive/1BBNNfBUNJPGT_7MxVboTqvRegJ-TUamc?usp=sharing

Here's why it didn't work for you:
  - PyMC implementation needs to output the upper triangular elements of the correlation matrix, where the TFP implementation outputs a Cholesky factor.
  - Differences in indexing off-diagonal elements. TFP actually fills in off-diagonal elements in a clockwise spiral, whereas np.triu_indices is row major. I notice you reverse the elements in the code above which is correct for 3x3 but not for larger matrices.

Reply:
I got access denied to the colab :)

I'm not surprised my code doesn't work, but I'm glad you know why. In the general case, tensorflow concatenates a reflected copy of the array to itself then reshapes and masks out the lower/upper triangle -- see [here ](https://www.tensorflow.org/probability/api_docs/python/tfp/math/fill_triangular)if you haven't already. 

I'm not sure that we need to copy their output 1:1 -- after all, the important thing is that we can go from unconstrained samples to a valid cholesky decomposed correlation matrix. Is the order we put the numbers into the matrix relevant? I'm not sure, but my instinct is no. On the other hand, if we copy 1:1 we can be sure it's right.

Reply:
> PyMC implementation needs to output the upper triangular elements of the correlation matrix, where the TFP implementation outputs a Cholesky factor.

Are you sure? I thought the upper triangular elements are the cholesky factorized correlation matrix. If you're right though we just need to add a matmul to the end right?

Reply:
Apologies - Link is now open with Viewer permission and I've made you an Editor.

I don't think the order we insert the off-diagonal elements into an array is very important, but it is needed in order to compare results between this implementation and the one in TFP. I would suggest sticking with `np.triu_indices` here.

Reply:
> Are you sure? I thought the upper triangular elements are the cholesky factorized correlation matrix. If you're right though we just need to add a matmul to the end right?

Yes, you can see this by looking at the implementation of LKJCorr. I originally thought the same thing, implemented the transform accordingly, then was surprised that non-posdef matrices were generated. :facepalm:

Reply:
Great! Sounds like you have a good handle on things. I think what would be a really important next step would be to add a test that your implementation correctly makes a round trip from $\mathbb R -> \Omega -> \mathbb R$, where $\Omega$ is the set of correlation matrices.

Reply:
Awesome - I can do that